### DISCUSSION TITLE
Conditional git_hash_ctx_cleanup call in git_odb_stream_free When stream->hash_ctx is set?

### DISCUSSION THREAD

**Discussion by gensmusic** [libgit2/libgit2#6835] [Q&A] [answered]
Hi,
I've been implementing custom odb backend with other language, `Rust`.

Odb backend needs to implenment two kinds of git_odb_stream, read and write.

When I create my `read stream`, l looks like this in Rust

```rust
raw::git_odb_stream {
      backend,
      mode: raw::GIT_STREAM_RDONLY,
      hash_ctx: ptr::null_mut(), // here is a null
      declared_size: 0,
      write: None,
      finalize_write: None,
      read: Some(git_odb_stream_read), // c callback
      free: Some(git_odb_stream_free), // c callback
      received_bytes: 0,
  },
```
But I eventully got an error from the floowing code in odb.c

```c
void git_odb_stream_free(git_odb_stream *stream)
{
	if (stream == NULL)
		return;

	git_hash_ctx_cleanup(stream->hash_ctx); // hash_ctx is null
	git__free(stream->hash_ctx);
	stream->free(stream);
}
```

I checked odb.c, and found out `stream->hash_ctx` is not used in read stream, and it's only used in write stream.
Therefore can we make git_hash_ctx_cleanup Conditionally called? like below

```c
void git_odb_stream_free(git_odb_stream *stream)
{
	if (stream == NULL)
		return;
        if (stream->hash_ctx)
	      git_hash_ctx_cleanup(stream->hash_ctx); 
	git__free(stream->hash_ctx);
	stream->free(stream);
}
```

If this is not allowed and hash_ctx is required anyway, I have to make a way to create a `git_hash_ctx` in Rust which requires a lot of work

**Comment by ethomson** [comment, marked as answer]
Hmm, really good question. This seems like we've definitely leaked out some core implementation details that shouldn't be something that consumers should be touching. read/write streams are not mandatory (libgit2 will wrap your normal, non-streaming read and write- functions with streaming proxies that will buffer). And because of that, I suspect that this oversight exists.

I think that the proper solution is for `git_odb_stream` to not actually have the `hash_ctx` member at all. (You'll run into the same problem if / when you try to build an odb write stream, I would expect?) But we can take your fix for a quick win while we untangle this.

**Reply by gensmusic** [reply]
> I think that the proper solution is for git_odb_stream to not actually have the hash_ctx member at all

Yes, that's better!

Should I make a PR or just wait for you guys to fix it? I really needs this fix.

**Reply by ethomson** [reply]
Please do send a PR with your proposed change (with the NULL check). I'm certainly happy if you send a PR for removing the `hash_ctx` -- but just to set expectations, I _think_ that removing `hash_ctx` is a bit of a gnarly change and will require some deeper refactoring.

**Comment by gensmusic** [comment]
@ethomson  A quick fix PR is sent https://github.com/libgit2/libgit2/pull/6836
